File: src/logger.c
/**
 * logger.c - Thread-safe file logging implementation
 */

#include "logger.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <bsd/string.h>
#include <stdarg.h>
#include <time.h>
#include <pthread.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>

// Global state
static FILE *g_log_file = NULL;
static LogLevel g_min_level = LOG_LEVEL_INFO;
static pthread_mutex_t g_log_mutex = PTHREAD_MUTEX_INITIALIZER;
static char g_log_path[512] = {0};
static long g_max_size_bytes = 10 * 1024 * 1024;  // 10 MB default
static int g_max_backups = 5;  // Keep 5 backup files
static char g_session_id[64] = {0};  // Session ID for log tagging
static int g_always_flush = 0;  // Default to buffered for better performance

// Level names for output
static const char *level_names[] = {
    "DEBUG",
    "INFO",
    "WARN",
    "ERROR"
};

/**
 * Create directory recursively (like mkdir -p)
 */
static int mkdir_p(const char *path) {
    char tmp[512];
    char *p = NULL;
    size_t len;

    snprintf(tmp, sizeof(tmp), "%s", path);
    len = strlen(tmp);

    // Remove trailing slash
    if (tmp[len - 1] == '/') {
        tmp[len - 1] = 0;
    }

    // Create directories recursively
    for (p = tmp + 1; *p; p++) {
        if (*p == '/') {
            *p = 0;
            if (mkdir(tmp, 0755) != 0 && errno != EEXIST) {
                return -1;
            }
            *p = '/';
        }
    }

    if (mkdir(tmp, 0755) != 0 && errno != EEXIST) {
        return -1;
    }

    return 0;
}

/**
 * Get default log file path
 * Priority:
 *   1. $CLAUDE_C_LOG_PATH (environment variable)
 *   2. $CLAUDE_C_LOG_DIR/claude.log (if CLAUDE_C_LOG_DIR is set)
 *   3. ./.claude-c/logs/claude.log (project-local)
 *   4. ~/.local/share/claude-c/logs/claude.log
 *   5. /tmp/claude-c.log (fallback)
 */
static int get_default_log_path(char *buffer, size_t buffer_size) {
    // Check for explicit log file path first
    const char *log_path_env = getenv("CLAUDE_C_LOG_PATH");
    if (log_path_env && log_path_env[0] != '\0') {
        snprintf(buffer, buffer_size, "%s", log_path_env);

        // Extract directory and ensure it exists
        char dir_path[512];
        snprintf(dir_path, sizeof(dir_path), "%s", log_path_env);
        char *last_slash = strrchr(dir_path, '/');
        if (last_slash) {
            *last_slash = '\0';
            if (mkdir_p(dir_path) != 0) {
                fprintf(stderr, "Warning: Failed to create log directory: %s\n", dir_path);
            }
        }

        return 0;
    }

    // Check for log directory override
    const char *log_dir_env = getenv("CLAUDE_C_LOG_DIR");
    if (log_dir_env && log_dir_env[0] != '\0') {
        if (mkdir_p(log_dir_env) == 0) {
            snprintf(buffer, buffer_size, "%s/claude.log", log_dir_env);
            return 0;
        } else {
            fprintf(stderr, "Warning: Failed to create log directory: %s\n", log_dir_env);
        }
    }

    // Try project-local .claude-c directory first
    if (mkdir_p("./.claude-c/logs") == 0) {
        snprintf(buffer, buffer_size, "./.claude-c/logs/claude.log");
        return 0;
    }

    const char *home = getenv("HOME");
    if (home) {
        // Try ~/.local/share/claude-c/logs/
        snprintf(buffer, buffer_size, "%s/.local/share/claude-c/logs", home);

        if (mkdir_p(buffer) == 0) {
            snprintf(buffer, buffer_size, "%s/.local/share/claude-c/logs/claude.log", home);
            return 0;
        }
    }

    // Fallback to /tmp
    snprintf(buffer, buffer_size, "/tmp/claude-c.log");
    return 0;
}

/**
 * Get current timestamp in ISO 8601 format
 */
static void get_timestamp(char *buffer, size_t buffer_size) {
    time_t now = time(NULL);
    struct tm *tm_info = localtime(&now);
    strftime(buffer, buffer_size, "%Y-%m-%d %H:%M:%S", tm_info);
}

/**
 * Extract just the filename from a full path
 */
static const char *get_filename(const char *path) {
    const char *last_slash = strrchr(path, '/');
    return last_slash ? last_slash + 1 : path;
}

/**
 * Get current log file size in bytes
 */
static long get_file_size(const char *path) {
    struct stat st;
    if (stat(path, &st) == 0) {
        return st.st_size;
    }
    return 0;
}

/**
 * Perform log rotation
 * Renames current log to .1, .1 to .2, etc.
 * Deletes oldest backup if max_backups is exceeded
 */
static void rotate_log(void) {
    char old_name[600];
    char new_name[600];

    // Close current log file
    if (g_log_file) {
        fclose(g_log_file);
        g_log_file = NULL;
    }

    // Delete oldest backup if it exists
    snprintf(old_name, sizeof(old_name), "%s.%d", g_log_path, g_max_backups);
    unlink(old_name);  // Ignore errors if file doesn't exist

    // Rotate existing backups: .N-1 -> .N
    for (int i = g_max_backups - 1; i >= 1; i--) {
        snprintf(old_name, sizeof(old_name), "%s.%d", g_log_path, i);
        snprintf(new_name, sizeof(new_name), "%s.%d", g_log_path, i + 1);
        rename(old_name, new_name);  // Ignore errors
    }

    // Rotate current log to .1
    snprintf(new_name, sizeof(new_name), "%s.1", g_log_path);
    rename(g_log_path, new_name);

    // Reopen log file (creates new empty file)
    g_log_file = fopen(g_log_path, "a");

    // Write rotation marker
    if (g_log_file) {
        char timestamp[64];
        get_timestamp(timestamp, sizeof(timestamp));
        fprintf(g_log_file, "=== Log rotated: %s ===\n", timestamp);
        fflush(g_log_file);
    }
}

/**
 * Check if rotation is needed and perform it
 * Called before each log write (already inside mutex)
 */
static void check_and_rotate(void) {
    if (!g_log_file || g_log_path[0] == '\0') {
        return;
    }

    long current_size = get_file_size(g_log_path);
    if (current_size >= g_max_size_bytes) {
        rotate_log();
    }
}

int log_init(void) {
    char log_path[512];

    if (get_default_log_path(log_path, sizeof(log_path)) != 0) {
        return -1;
    }

    // Check for flush mode environment variable
    const char *flush_mode_env = getenv("CLAUDE_LOG_FLUSH");
    if (flush_mode_env && strcmp(flush_mode_env, "always") == 0) {
        g_always_flush = 1;
    } else {
        g_always_flush = 0;  // Default to buffered for better performance
    }

    return log_init_with_path(log_path);
}

int log_init_with_path(const char *log_path) {
    pthread_mutex_lock(&g_log_mutex);

    // Close existing log file if open
    if (g_log_file) {
        fclose(g_log_file);
        g_log_file = NULL;
    }

    // Open log file in append mode
    g_log_file = fopen(log_path, "a");
    if (!g_log_file) {
        pthread_mutex_unlock(&g_log_mutex);
        fprintf(stderr, "Failed to open log file: %s\n", log_path);
        return -1;
    }

    // Store path for reference
    strlcpy(g_log_path, log_path, sizeof(g_log_path));

    // Write startup marker
    char timestamp[64];
    get_timestamp(timestamp, sizeof(timestamp));
    fprintf(g_log_file, "\n=== Log started: %s (PID: %d) ===\n", timestamp, getpid());
    fflush(g_log_file);

    pthread_mutex_unlock(&g_log_mutex);
    return 0;
}

void log_set_level(LogLevel level) {
    pthread_mutex_lock(&g_log_mutex);
    g_min_level = level;
    pthread_mutex_unlock(&g_log_mutex);
}

void log_set_rotation(int max_size_mb, int max_backups) {
    pthread_mutex_lock(&g_log_mutex);
    g_max_size_bytes = (long)max_size_mb * 1024 * 1024;
    g_max_backups = max_backups;
    pthread_mutex_unlock(&g_log_mutex);
}

void log_set_session_id(const char *session_id) {
    pthread_mutex_lock(&g_log_mutex);
    if (session_id) {
        strlcpy(g_session_id, session_id, sizeof(g_session_id));
    } else {
        g_session_id[0] = '\0';
    }
    pthread_mutex_unlock(&g_log_mutex);
}

void log_set_flush_mode(int always_flush) {
    pthread_mutex_lock(&g_log_mutex);
    g_always_flush = always_flush;
    pthread_mutex_unlock(&g_log_mutex);
}

void log_message(LogLevel level, const char *file, int line,
                const char *func, const char *fmt, ...) {
    // Quick check without lock for performance
    if (level < g_min_level) {
        return;
    }

    pthread_mutex_lock(&g_log_mutex);

    if (!g_log_file) {
        pthread_mutex_unlock(&g_log_mutex);
        return;
    }

    // Check if rotation is needed
    check_and_rotate();

    if (!g_log_file) {  // Rotation might have failed
        pthread_mutex_unlock(&g_log_mutex);
        return;
    }

    // Get timestamp
    char timestamp[64];
    get_timestamp(timestamp, sizeof(timestamp));

    // Format: [TIMESTAMP] [SESSION_ID] LEVEL [file:line] function: message
    if (g_session_id[0] != '\0') {
        fprintf(g_log_file, "[%s] [%s] %-5s [%s:%d] %s: ",
                timestamp,
                g_session_id,
                level_names[level],
                get_filename(file),
                line,
                func);
    } else {
        // Fallback format without session ID
        fprintf(g_log_file, "[%s] %-5s [%s:%d] %s: ",
                timestamp,
                level_names[level],
                get_filename(file),
                line,
                func);
    }

    // Write the actual log message
    va_list args;
    va_start(args, fmt);
    vfprintf(g_log_file, fmt, args);
    va_end(args);

    fprintf(g_log_file, "\n");

    // Flush based on configured mode
    if (g_always_flush || level >= LOG_LEVEL_WARN) {
        fflush(g_log_file);
    }

    pthread_mutex_unlock(&g_log_mutex);
}

void log_flush(void) {
    pthread_mutex_lock(&g_log_mutex);

    if (g_log_file) {
        fflush(g_log_file);
    }

    pthread_mutex_unlock(&g_log_mutex);
}

void log_shutdown(void) {
    pthread_mutex_lock(&g_log_mutex);

    if (g_log_file) {
        char timestamp[64];
        get_timestamp(timestamp, sizeof(timestamp));
        fprintf(g_log_file, "=== Log ended: %s ===\n\n", timestamp);
        fflush(g_log_file);
        fclose(g_log_file);
        g_log_file = NULL;
    }

    pthread_mutex_unlock(&g_log_mutex);
}

File: zig/logger.zig
//! Logger — Thread-safe file logging
//!
//! Idiomatic Zig replacement for src/logger.c
//!
//! Key C→Zig translations:
//!   - `pthread_mutex_t`  → `std.Thread.Mutex`
//!   - `va_list` varargs  → comptime `fmt` + `anytype args` (std.fmt.format)
//!   - `rotate_log`       → `Logger.rotate` (same logic, explicit ownership)
//!   - Global state       → `Logger` struct (explicit, thread-safe via Mutex)
//!   - Macros (LOG_INFO)  → `logger.log(.info, ...)` (no macros needed)
//!
//! ## Usage
//!
//! ```zig
//! var logger = try Logger.init(std.heap.page_allocator, null); // null = auto-detect path
//! defer logger.deinit();
//! logger.log(.info, "started PID={d}", .{pid});
//! logger.setLevel(.warn); // suppress debug/info
//! ```
//!
//! The default log path follows the same priority as the C implementation:
//!   1. `$KLAWED_LOG_PATH`
//!   2. `$KLAWED_LOG_DIR/klawed.log`
//!   3. `./.klawed/logs/klawed.log`
//!   4. `~/.local/share/klawed/logs/klawed.log`
//!   5. `/tmp/klawed.log`

const std = @import("std");

// ---------------------------------------------------------------------------
// Log level
// ---------------------------------------------------------------------------

pub const Level = enum(u8) {
    fine = 0,
    debug = 1,
    info = 2,
    warn = 3,
    err = 4,

    pub fn name(self: Level) []const u8 {
        return switch (self) {
            .fine => "FINE ",
            .debug => "DEBUG",
            .info => "INFO ",
            .warn => "WARN ",
            .err => "ERROR",
        };
    }
};

// ---------------------------------------------------------------------------
// Logger struct
// ---------------------------------------------------------------------------

pub const Logger = struct {
    allocator: std.mem.Allocator,
    file: ?std.fs.File,
    path: []u8,
    min_level: Level,
    mutex: std.Thread.Mutex,
    max_size_bytes: u64,
    max_backups: u32,
    session_id: [64]u8,
    session_id_len: usize,
    always_flush: bool,

    /// Open (or create) a log file.
    /// Pass `null` for `log_path` to auto-detect using environment variables
    /// and standard fallback directories.
    pub fn init(allocator: std.mem.Allocator, log_path: ?[]const u8) !Logger {
        const path = if (log_path) |p|
            try allocator.dupe(u8, p)
        else
            try detectLogPath(allocator);

        // Ensure parent directory exists
        if (std.fs.path.dirname(path)) |dir| {
            std.fs.cwd().makePath(dir) catch {}; // best effort
        }

        const file = std.fs.createFileAbsolute(path, .{ .truncate = false }) catch |err| {
            std.debug.print("logger: failed to open {s}: {}\n", .{ path, err });
            return err;
        };
        // Seek to end for append behaviour
        try file.seekFromEnd(0);

        return Logger{
            .allocator = allocator,
            .file = file,
            .path = path,
            .min_level = .info,
            .mutex = .{},
            .max_size_bytes = 10 * 1024 * 1024, // 10 MiB
            .max_backups = 5,
            .session_id = undefined,
            .session_id_len = 0,
            .always_flush = false,
        };
    }

    /// Close the log file and free internal memory.
    pub fn deinit(self: *Logger) void {
        if (self.file) |f| {
            self.writeMarker("Log ended") catch {};
            f.close();
            self.file = null;
        }
        self.allocator.free(self.path);
    }

    /// Set the minimum log level.  Messages below this level are ignored.
    pub fn setLevel(self: *Logger, level: Level) void {
        self.mutex.lock();
        defer self.mutex.unlock();
        self.min_level = level;
    }

    /// Set the session ID included in every log line.
    pub fn setSessionId(self: *Logger, id: []const u8) void {
        self.mutex.lock();
        defer self.mutex.unlock();
        const n = @min(id.len, self.session_id.len - 1);
        @memcpy(self.session_id[0..n], id[0..n]);
        self.session_id[n] = 0;
        self.session_id_len = n;
    }

    /// Configure log rotation.
    pub fn setRotation(self: *Logger, max_size_mb: u32, max_backups: u32) void {
        self.mutex.lock();
        defer self.mutex.unlock();
        self.max_size_bytes = @as(u64, max_size_mb) * 1024 * 1024;
        self.max_backups = max_backups;
    }

    /// Set whether to flush after every write.
    pub fn setAlwaysFlush(self: *Logger, always: bool) void {
        self.mutex.lock();
        defer self.mutex.unlock();
        self.always_flush = always;
    }

    /// Log a message at `level` using `std.fmt` formatting.
    pub fn log(
        self: *Logger,
        level: Level,
        comptime fmt: []const u8,
        args: anytype,
    ) void {
        if (@intFromEnum(level) < @intFromEnum(self.min_level)) return;

        self.mutex.lock();
        defer self.mutex.unlock();

        const f = self.file orelse return;

        self.maybeRotate() catch {};

        const writer = f.writer();

        // Timestamp
        const ts = timestampNow();
        writer.print("[{s}] ", .{&ts}) catch return;

        // Optional session ID
        if (self.session_id_len > 0) {
            writer.print("[{s}] ", .{self.session_id[0..self.session_id_len]}) catch return;
        }

        // Level + message
        writer.print("{s} ", .{level.name()}) catch return;
        writer.print(fmt ++ "\n", args) catch return;

        if (self.always_flush or @intFromEnum(level) >= @intFromEnum(Level.warn)) {
            f.sync() catch {};
        }
    }

    /// Flush buffered data to disk.
    pub fn flush(self: *Logger) void {
        self.mutex.lock();
        defer self.mutex.unlock();
        if (self.file) |f| f.sync() catch {};
    }

    // -----------------------------------------------------------------------
    // Private helpers
    // -----------------------------------------------------------------------

    fn writeMarker(self: *Logger, msg: []const u8) !void {
        const f = self.file orelse return;
        const ts = timestampNow();
        try f.writer().print("=== {s}: {s} ===\n", .{ msg, &ts });
    }

    fn maybeRotate(self: *Logger) !void {
        const f = self.file orelse return;
        const size = (f.getPos() catch 0);
        if (size < self.max_size_bytes) return;

        // Close current file
        f.close();
        self.file = null;

        // Delete oldest backup
        const oldest = try std.fmt.allocPrint(
            self.allocator,
            "{s}.{d}",
            .{ self.path, self.max_backups },
        );
        defer self.allocator.free(oldest);
        std.fs.deleteFileAbsolute(oldest) catch {};

        // Rotate .N-1 → .N
        var i: u32 = self.max_backups;
        while (i > 1) : (i -= 1) {
            const old = try std.fmt.allocPrint(self.allocator, "{s}.{d}", .{ self.path, i - 1 });
            defer self.allocator.free(old);
            const new = try std.fmt.allocPrint(self.allocator, "{s}.{d}", .{ self.path, i });
            defer self.allocator.free(new);
            std.fs.renameAbsolute(old, new) catch {};
        }

        // Rotate current → .1
        const dot1 = try std.fmt.allocPrint(self.allocator, "{s}.1", .{self.path});
        defer self.allocator.free(dot1);
        std.fs.renameAbsolute(self.path, dot1) catch {};

        // Reopen
        const new_file = try std.fs.createFileAbsolute(self.path, .{ .truncate = true });
        self.file = new_file;

        const ts = timestampNow();
        new_file.writer().print("=== Log rotated: {s} ===\n", .{&ts}) catch {};
    }
};

// ---------------------------------------------------------------------------
// Timestamp helper
// ---------------------------------------------------------------------------

/// Return a 19-char "YYYY-MM-DD HH:MM:SS" timestamp without allocation.
fn timestampNow() [19]u8 {
    const c = @cImport(@cInclude("time.h"));
    var ts: c.time_t = c.time(null);
    var tm_buf: c.struct_tm = undefined;
    _ = c.localtime_r(&ts, &tm_buf);
    var buf: [19]u8 = undefined;
    _ = std.fmt.bufPrint(&buf, "{d:0>4}-{d:0>2}-{d:0>2} {d:0>2}:{d:0>2}:{d:0>2}", .{
        @as(u32, @intCast(tm_buf.tm_year + 1900)),
        @as(u32, @intCast(tm_buf.tm_mon + 1)),
        @as(u32, @intCast(tm_buf.tm_mday)),
        @as(u32, @intCast(tm_buf.tm_hour)),
        @as(u32, @intCast(tm_buf.tm_min)),
        @as(u32, @intCast(tm_buf.tm_sec)),
    }) catch unreachable;
    return buf;
}

// ---------------------------------------------------------------------------
// Log path detection
// ---------------------------------------------------------------------------

/// Detect the default log path using the same priority as the C logger.
/// Caller must free the returned path with `allocator.free`.
fn detectLogPath(allocator: std.mem.Allocator) ![]u8 {
    // 1. $KLAWED_LOG_PATH
    if (std.process.getEnvVarOwned(allocator, "KLAWED_LOG_PATH") catch null) |p| {
        return p;
    }

    // 2. $KLAWED_LOG_DIR/klawed.log
    if (std.process.getEnvVarOwned(allocator, "KLAWED_LOG_DIR") catch null) |dir| {
        defer allocator.free(dir);
        std.fs.cwd().makePath(dir) catch {};
        return std.fs.path.join(allocator, &.{ dir, "klawed.log" });
    }

    // 3. ./.klawed/logs/klawed.log
    {
        const local = ".klawed/logs";
        if (std.fs.cwd().makePath(local)) |_| {
            const cwd_path = try std.fs.realpathAlloc(allocator, ".");
            defer allocator.free(cwd_path);
            return std.fs.path.join(allocator, &.{ cwd_path, ".klawed/logs/klawed.log" });
        } else |_| {}
    }

    // 4. ~/.local/share/klawed/logs/klawed.log
    if (std.process.getEnvVarOwned(allocator, "HOME") catch null) |home| {
        defer allocator.free(home);
        const dir = try std.fs.path.join(allocator, &.{ home, ".local/share/klawed/logs" });
        defer allocator.free(dir);
        if (std.fs.cwd().makePath(dir)) |_| {
            return std.fs.path.join(allocator, &.{ dir, "klawed.log" });
        } else |_| {}
    }

    // 5. /tmp/klawed.log
    return allocator.dupe(u8, "/tmp/klawed.log");
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "Logger: init, log, deinit" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const log_path = try std.fs.path.join(std.testing.allocator, &.{ tmp_path, "test.log" });
    defer std.testing.allocator.free(log_path);

    var logger = try Logger.init(std.testing.allocator, log_path);
    defer logger.deinit();

    logger.log(.info, "hello {s}", .{"world"});
    logger.log(.warn, "warning {d}", .{42});
    logger.flush();

    // Verify the log file has content
    const contents = try std.fs.File.readToEndAlloc(
        try std.fs.openFileAbsolute(log_path, .{}),
        std.testing.allocator,
        1024 * 1024,
    );
    defer std.testing.allocator.free(contents);

    try std.testing.expect(std.mem.indexOf(u8, contents, "hello world") != null);
    try std.testing.expect(std.mem.indexOf(u8, contents, "warning 42") != null);
}

test "Logger: level filtering" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const log_path = try std.fs.path.join(std.testing.allocator, &.{ tmp_path, "filter.log" });
    defer std.testing.allocator.free(log_path);

    var logger = try Logger.init(std.testing.allocator, log_path);
    defer logger.deinit();

    logger.setLevel(.err); // only errors
    logger.log(.info, "this should NOT appear", .{});
    logger.log(.err, "this SHOULD appear", .{});
    logger.flush();

    const contents = try std.fs.File.readToEndAlloc(
        try std.fs.openFileAbsolute(log_path, .{}),
        std.testing.allocator,
        1024 * 1024,
    );
    defer std.testing.allocator.free(contents);

    try std.testing.expect(std.mem.indexOf(u8, contents, "should NOT appear") == null);
    try std.testing.expect(std.mem.indexOf(u8, contents, "this SHOULD appear") != null);
}

test "Logger: session ID included in output" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const log_path = try std.fs.path.join(std.testing.allocator, &.{ tmp_path, "sess.log" });
    defer std.testing.allocator.free(log_path);

    var logger = try Logger.init(std.testing.allocator, log_path);
    defer logger.deinit();

    logger.setSessionId("sess_abc123");
    logger.log(.info, "with session", .{});
    logger.flush();

    const contents = try std.fs.File.readToEndAlloc(
        try std.fs.openFileAbsolute(log_path, .{}),
        std.testing.allocator,
        1024 * 1024,
    );
    defer std.testing.allocator.free(contents);

    try std.testing.expect(std.mem.indexOf(u8, contents, "sess_abc123") != null);
}

test "Level.name: returns correct strings" {
    try std.testing.expectEqualStrings("INFO ", Level.info.name());
    try std.testing.expectEqualStrings("WARN ", Level.warn.name());
    try std.testing.expectEqualStrings("ERROR", Level.err.name());
    try std.testing.expectEqualStrings("DEBUG", Level.debug.name());
    try std.testing.expectEqualStrings("FINE ", Level.fine.name());
}
